RFC: Local snapshot file format (#25)
iotaledger/protocol-rfcs: text/0000-local-snapshot-file-format/0000-local-snapshot-file-format.md

+ Feature name: `local_snapshot_file_format`
+ Start date: 2020-08-25
+ RFC PR: [iotaledger/protocol-rfcs#0000](https://github.com/iotaledger/protocol-rfcs/pull/0000)

# Summary

This RFC defines a file format for local snapshots which is compatible with Chrysalis Phase 2.

# Motivation

> Nodes create local snapshots to produce ledger representations at a point in time of a given milestone in order to be able to:
> * start up from a recent milestone instead of having to synchronize from the genesis transaction.
> * delete old transaction data below a given milestone.
> 
> Current node implementations use a [local snapshot file format](https://github.com/iotaledger/iri-ls-sa-merger/tree/351020d3b5e342b6e9a41f2868575ab7ff8c251c#generating-an-export-file-from-a-localsnapshots-db) which only works with account based ledgers. For Chrysalis Phase 2 this file format has to be assimilated to support a UTXO based ledger.

Reviewer: ```suggestion
Nodes create local snapshots to produce ledger representations at a point in time of a given milestone to be able to:
* Start up from a recent milestone instead of having to synchronize from the genesis transaction.
* Delete old transaction data below a given milestone.

Current node implementations use a [local snapshot file format](https://github.com/iotaledger/iri-ls-sa-merger/tree/351020d3b5e342b6e9a41f2868575ab7ff8c251c#generating-an-export-file-from-a-localsnapshots-db) which only works with account based ledgers. For Chrysalis Phase 2, this file format has to be assimilated to support a UTXO based ledger.
```

# Detailed design

Since a UTXO based ledger is much larger in size, this RFC proposes two formats for snapshot files: 
* A `full` format which represents a complete ledger state.

> * A `delta` format which only contains diffs (consumed and spent outputs) of milestones from a given milestone index onwards.

Reviewer 2: > consumed and spent

So far the only output type that can be consumed is a TX, which implies spending those funds. I would remove the "consumed" term as it implies that are already outputs that can be consumed without spending: such as Coordicie's mana pledges.

This separation allows nodes to swiftly create new delta snapshot files, which then can be distributed with a companion full snapshot file to reconstruct a recent state.

> Unlike the current format, these new formats do not include spent addresses since this information is no longer held by nodes.

Reviewer 2: The reason for removing them is the discontinuing of WOTS (if we finally come to an agreement in that regard).

> # Detailed design
> 
> Since a UTXO based ledger is much larger in size, this RFC proposes two formats for snapshot files: 
> * a `full` format which represents a complete ledger state.
> * a `delta` format which only contains diffs (consumed and spent outputs) of milestones from a given milestone index onwards.
> 
> This separation allows nodes to swiftly create new delta snapshot files, which then can be distributed with a companion full snapshot file to reconstruct a recent state.
> 
> Unlike the current format, this new formats do not longer include spent addresses since this information is no longer held by nodes.

Reviewer: ```suggestion
# Detailed design

Since a UTXO based ledger is much larger in size, this RFC proposes two formats for snapshot files: 
* A `full` format which represents a complete ledger state.
* A `delta` format which only contains diffs (consumed and spent outputs) of milestones from a given milestone index onwards.

This separation allows nodes to swiftly create new delta snapshot files, which then can be distributed with a companion full snapshot file to reconstruct a recent state.

Unlike the current format, these new formats do not include spent addresses since this information is no longer held by nodes.
```

### Formats

> All types are serialized in little-endian

#### Full Ledger State

A full ledger snapshot file contains the UTXOs (`outputs` section) of a node's latest solid milestone (`ledger_milestone_index`). The `diffs` contain the diffs to rollback the `outputs` state to regain the ledger state of the snapshot milestone at (`seps_milestone_index`).

![](https://i.imgur.com/e6WuufK.png)

> While the node producing such a full ledger state snapshot could theoretically pre-compute the actual snapshot milestone state, this is deferred to the consumer of the data to speed up local snapshot creation.

Reviewer 2: I am starting to get a bit confused about the terminology: "actual snapshot milestone" == "seps_miletone_index"?

> A full ledger snapshot file contains the UTXOs (`outputs` section) of a node's latest solid milestone (`ledger_milestone_index`). The `diffs` contain the diffs to rollback the `outputs` state in order to regain the ledger state of the snapshot milestone at (`seps_milestone_index`).
> 
> ![](https://i.imgur.com/e6WuufK.png)
> 
> While the node producing such full ledger state snapshot could theoretically pre-compute the actual snapshot milestone state, this is deferred to the consumer of the data in order to speed up local snapshot creation.
> 
> A full ledger state local snapshot is denoted by the type byte `0`:

Reviewer: ```suggestion
A full ledger snapshot file contains the UTXOs (`outputs` section) of a node's latest solid milestone (`ledger_milestone_index`). The `diffs` contain the diffs to rollback the `outputs` state to regain the ledger state of the snapshot milestone at (`seps_milestone_index`).

![](https://i.imgur.com/e6WuufK.png)

While the node producing such a full ledger state snapshot could theoretically pre-compute the actual snapshot milestone state, this is deferred to the consumer of the data to speed up local snapshot creation.

A full ledger state local snapshot is denoted by the type byte `0`:
```

```
version<byte>
type<byte> = 0
timestamp<uint64>
network_id<uint64>
seps_milestone_index<uint64>
ledger_milestone_index<uint64>
seps_count<uint64>
outputs_count<uint64>
diffs_count<uint64>
seps<array[seps_count]>:
	sep<array[32]>
outputs<array[outputs_count]>:

> 	transaction_hash<array[32]>

Reviewer 3: Name is confusing me, is this supposed to be a transaction id ? Not a message id ?

	output_index<uint16>
	address:
		address_type<byte>
		wots_address<array[49]>
		||
		address_type<byte>
		ed25519_address<array[32]>
	value<uint64>

> diffs<array[diffs_count]>:

Reviewer 2: There is no specific ordering of diffs you enforce here. You can have a diff for milestone X followed by a diff for milestone X+100. For integrity and validation sake it would be nicer to have an `array[ledger_milestone_index - seps_milestone_index]` of an array of diffs between milestone X and X+1, sorted by X, starting from `seps_milestone_index` and ending on `ledger_milestone_index`.

A delta ledger state local snapshot only contains the `diffs` of milestones starting from a given `ledger_milestone_index`. A node consuming such data must know the state of the ledger at `ledger_milestone_index`.

![](https://i.imgur.com/bt5BUpe.png)

A delta ledger state local snapshot is denoted by the type byte `1`:

```
version<byte>
type<byte> = 1
timestamp<uint64>
network_id<uint64>
seps_milestone_index<uint64>
ledger_milestone_index<uint64>
seps_count<uint64>
diffs_count<uint64>
seps<array[seps_count]>:
	sep<array[32]>
diffs<array[diffs_count]>:
	diff_milestone_index<array[32]>
	created_outputs_count<uint64>
	created_outputs<array>:
		transaction_hash<array[32]>
		output_index<uint16>
		address:
			address_type<byte> = 0
			wots_address<array[49]>
			||
			address_type<byte> = 1
			ed25519_address<array[32]>
		value<uint64>
	consumed_outputs_count<uint64>
	consumed_outputs<array>:
		transaction_hash<array[32]>
		output_index<uint16>
		address:
			address_type<byte> = 0
			wots_address<array[49]>
			||
			address_type<byte> = 1
			ed25519_address<array[32]>
		value<uint64>
```

# Drawbacks

Nodes need to support this new format.

# Rationale and alternatives

> * In conjunction with a companion full snapshot, a tool or node can "truncate" the data from a delta snapshot back to a single full snapshot. In that case, the `ledger_milestone_index` and `seps_milestone_index` would be the same. In the example above, given the full and delta snapshots, one could produce a new full snapshot for milestone 1350.

Reviewer 2: Here you mean that, after the truncation, the `seps_milestone_index` of the Delta will be == to the  `ledger_milestone_index` of the Full. Correct?

* Since snapshots may include millions of UTXOs, code generating such files needs to stream data directly onto disk instead of keeping the entire representation in memory. In order to facilitate this, the count denotations for SEPs, UTXOs and diffs are at the beginning of the file. This allows code generating snapshot files to only have to seek back once after the actual count of elements is known.

# Unresolved questions

* Is all the information to startup a node from the local snapshot available with the described format?
* Can we get rid of the spent addresses or do we still need to keep them?

> * Do we need to account for different types of outputs already? (we currently only have them deposit to addresses)

Reviewer 2: I am afraid so... https://github.com/iotaledger/protocol-rfcs/pull/32/ already introduces a new output type that is crucial for validation.

### Formats

> All types are serialized in little-endian

#### Full Ledger State

A full ledger snapshot file contains the UTXOs (`outputs` section) of a node's latest solid milestone (`ledger_milestone_index`). The `diffs` contain the diffs to rollback the `outputs` state in order to regain the ledger state of the snapshot milestone at (`seps_milestone_index`).

![](https://i.imgur.com/e6WuufK.png)

While the node producing such full ledger state snapshot could theoretically pre-compute the actual snapshot milestone state, this is deferred to the consumer of the data in order to speed up local snapshot creation.

A full ledger state local snapshot is denoted by the type byte `0`:

```
version<byte>
type<byte> = 0
timestamp<uint64>
seps_milestone_index<uint64>
seps_milestone_hash<array[32]>
ledger_milestone_index<uint64>
ledger_milestone_hash<array[32]>
seps_count<uint64>
outputs_count<uint64>
diffs_count<uint64>
seps<array[seps_count]>:
	sep<array[32]>
outputs<array[outputs_count]>:
	transaction_hash<array[32]>
	output_index<uint16>
	address:
		address_type<byte>
		wots_address<array[49]>
		||

> 		address_tybe<byte>

Reviewer 3: ```suggestion
		address_type<byte>
```

		ed25519_address<array[32]>
	value<uint64>
diffs<array[diffs_count]>:
	diff_milestone_index<array[32]>
	created_outputs_count<uint64>
	created_outputs<array>:
		transaction_hash<array[32]>
		output_index<uint16>
		address:
			address_type<byte> = 0
			wots_address<array[49]>
			||

> 			address_tybe<byte> = 1

Reviewer 3: ```suggestion
			address_type<byte> = 1
```

			ed25519_address<array[32]>
		value<uint64>
	consumed_outputs_count<uint64>
	consumed_outputs<array>:
		transaction_hash<array[32]>
		output_index<uint16>
		address:
			address_type<byte>
			wots_address<array[49]>
			||

> 			address_tybe<byte>

Reviewer 3: ```suggestion
			address_type<byte>
```

			ed25519_address<array[32]>
		value<uint64>		
```

#### Delta Ledger State

A delta ledger state local snapshot only contains the `diffs` of milestones starting from a given `ledger_milestone_index`. A node consuming such data must know the state of the ledger at `ledger_milestone_index`.

![](https://i.imgur.com/bt5BUpe.png)

A delta ledger state local snapshot is denoted by the type byte `1`:

> 

Reviewer 4: Even though I agree a node should keep track of the diff at every milestone, should the snapshot file keep track of it?
I would just expect a diff from the base

Reviewer 2: I think it makes sense to include information at every milestone actually: especially given the fact that it will be a good thing to introduce MS data as well (proof of inclusion specifically).

```
version<byte>
type<byte> = 1
timestamp<uint64>
seps_milestone_index<uint64>
seps_milestone_hash<array[32]>
ledger_milestone_index<uint64>
ledger_milestone_hash<array[32]>
seps_count<uint64>
diffs_count<uint64>
seps<array[seps_count]>:
	sep<array[32]>
diffs<array[diffs_count]>:
	diff_milestone_index<array[32]>
	created_outputs_count<uint64>
	created_outputs<array>:
		transaction_hash<array[32]>
		output_index<uint16>
		address:
			address_type<byte> = 0
			wots_address<array[49]>
			||

> 			address_tybe<byte> = 1

Reviewer 3: ```suggestion
			address_type<byte> = 1
```

			ed25519_address<array[32]>
		value<uint64>
	consumed_outputs_count<uint64>
	consumed_outputs<array>:
		transaction_hash<array[32]>
		output_index<uint16>
		address:
			address_type<byte> = 0
			wots_address<array[49]>
			||

> 			address_tybe<byte> = 1

Reviewer 3: ```suggestion
			address_type<byte> = 1
```

			ed25519_address<array[32]>
		value<uint64>
```

# Drawbacks

> Nodes need to support this new format.

Reviewer 4: I guess you had nothing to fill in :-) ?

Maybe:
1. We have 2 different snapshot files which may be confusing
2. The size of the file will grow as the number of outputs increase

+ Feature name: `local_snapshot_file_format`
+ Start date: 2020-08-25
+ RFC PR: [iotaledger/protocol-rfcs#0000](https://github.com/iotaledger/protocol-rfcs/pull/0000)

# Summary

This RFC defines a file format for local snapshots which is compatible with Chrysalis Phase 2.

# Motivation

Nodes create local snapshots to produce ledger representations at a point in time of a given milestone in order to be able to:
* start up from a recent milestone instead of having to synchronize from genesis
* delete transaction data below the given milestone

Current node implementations use a [local snapshot file format](https://github.com/iotaledger/iri-ls-sa-merger/tree/351020d3b5e342b6e9a41f2868575ab7ff8c251c#generating-an-export-file-from-a-localsnapshots-db) which only works with account based ledgers. For Chrysalis Phase 2 this file format has to be assimilated to support a UTXO based ledger.

# Detailed design

All types are serialized in little-endian and occur in the sequence of the rows defined below. Local snapshot files are compressed via zlib to further reduce size.

> SEP = solid entry point. `Array[T]` are prefixed with a varint denoting the length.

Reviewer 4: maybe add a one liner explanation of what is an SEP?

> SEP = solid entry point. `Array[T]` are prefixed with a varint denoting the length.

Reviewer 2: ```suggestion
SEP = solid entry point. `Array[T]` are prefixed with a varint denoting the array element count.
```

Author: Please look at the new changes.

This format describes version 1:
<table>
    <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
    </tr>
    <tr>
        <td>Version</td>

>         <td>byte</td>

Reviewer 2: Version fields are specified as `varint`s in all other RFCs.

        <td>
        The version of the local snapshot file format.
        </td>
    </tr>    
    <tr>
        <td>Milestone Index</td>
        <td>uint64</td>
        <td>
        The index of the milestone of the local snapshot.
        </td>
    </tr>
    <tr>
        <td>Milestone Hash</td>
        <td>ByteArray[32]</td>
        <td>
        The BLAKE2b-256 hash of the milestone payload.
        </td>
    </tr>
    <tr>
        <td>Timestamp</td>
        <td>uint64</td>
        <td>
        The UNIX epoch timestamp in seconds of when this snapshot was created.
        </td>
    </tr>
    <tr>
        <td>SEPs</td>
        <td>Array[ByteArray[32]]</td>
        <td>
        The BLAKE2b-256 hashes of the SEP messages at the cut off point of the given milestone.
        </td>
    </tr>
    <tr>
        <td>UTXOs</td>
        <td colspan="2">
            <details open="true">

>                 <summary>Array[UTXO]</summary>

Reviewer 2: If I understand correctly this array will be preceded by a varint specifying the element count of the Array itself; while forming the file you may not know ahead of time how many UTXOs to persist, and therefore the precise offset of the UTXOs array, shifted by its count varint. Seeking and moving on that file descriptior might be quite annoying: perhaps it would be better to detach your Array definition from its respective varint count, so you can manipulate the type of this latter depending on the circumstance. In this case you are probably better off with a fixed type for the UTXOs count.

Author: Yeah unfortunately I only wrote the export code after writing the RFC. I am currently writing up another format which also allows iterators for UTXOs.

                <blockquote>
                Describes the unspent transaction outputs per transaction.
                </blockquote>
                <table>
                    <tr>
                        <td><b>Name<b></td>
                        <td><b>Type</b></td>
                        <td><b>Description</b></td>
                    </tr>
                    <tr>
                        <td>Transaction Hash</td>
                        <td>ByteArray[32]</td>
                        <td>The BLAKE2b-256 hash of the transaction.</td>
                    </tr>
                    <tr>
                        <td>
                        Unspent Outputs
                        </td>
                        <td colspan="2">
                            <details open="true">
                                <summary>Array[Outputs]</summary>
                                <table>
                                    <tr>
                                        <td>Index</td>

>                                         <td>byte</td>

Reviewer 2: This should be a `varint` to be consistent with [Inputs / Outputs Count fields](https://github.com/luca-moser/protocol-rfcs/blob/signed-tx-payload/text/0000-signed-transaction-payload/0000-signed-transaction-payload.md#serialized-layout), even if we "informally" define such count to be < 127.

Reviewer 5: I guess this needs to be revisited once, when exactly that part of RFC #18 has been finalized. `byte` definitely seems to restricting to me.

Author: All array lengths will be replaced with fixed sized integers.

                                                    <tr>
                                                        <td>Address</td>
                                                        <td>ByteArray[32]</td>
                                                        <td>The raw bytes of the Ed25519 address which is a BLAKE2b-256 hash of the Ed25519 public key.</td>
                                                    </tr>
                                                </table>
                                            </details>
                                        </td>
                                    </tr>
                                    <tr>
                                        <td>Value</td>
                                        <td>uint64</td>
                                        <td>The output value.</td>
                                    </tr>
                                </table>
                            </details>
                        </td>
                    </tr>
                </table>
            </details>
        </td>
    </tr>
    <tr>
        <td>Integrity hash</td>
        <td>ByteArray[32]</td>
        <td>

>         The SHA256 hash of all the previous fields.

Reviewer 2: Why introducing another hash type instead of using `BLAKE2b-256`?

Reviewer 5: I was wondering about the same thing :thinking: 
What is actually the purpose of this?

If it's about authenticity it should be a signature, shouldn't it? 
If it is just about data integrity, I think this should be part of the second layer. Exactly like the compression.
Probably, I would just remove it and lift that functionality higher.

Author: Will be removed.

            ed25519_address<array[32]>
            ||
            ...
        value<uint64>
        ||
        output_type<byte> = 1 // sig_locked_dust_allowance_output
        address:
            address_type<byte> = 1
            ed25519_address<array[32]>
            ||
            ...
        value<uint64>
diffs<array[diffs_count]>:
    diff_milestone_index<array[32]>
    created_outputs_count<uint64>
    created_outputs<array>:
        message_hash<array[32]>
        transaction_hash<array[32]>
        output_index<uint16>
        output:
            output_type<byte> = 0 // sig_locked_single_output
            address:
                address_type<byte> = 1
                ed25519_address<array[32]>
                ||
                ...
            value<uint64>
            ||
            output_type<byte> = 1 // sig_locked_dust_allowance_output
            address:
                address_type<byte> = 1
                ed25519_address<array[32]>
                ||
                ...
            value<uint64>
    consumed_outputs_count<uint64>
    consumed_outputs<array>:
        message_hash<array[32]>
        transaction_hash<array[32]>

>         output_index<uint16>

Reviewer 3: I think this is missing the target TransactionId.
